### PR TITLE
Cook README for Tide

### PR DESCRIPTION
**Pull request by ingvagabund**
Fixes: https://github.com/kubernetes/test-infra/issues/5919

### CODE REVIEW

File: prow/tide/README.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
+# Tide
+
+One of [Prow](https://github.com/kubernetes/test-infra/blob/master/prow/README.md)
```
Comment (cjwagner) [inline prow/tide/README.md]: s/One of/Tide is a/

We don't have any other Prow components that do this.

File: prow/tide/README.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+components for managing a pool of PRs that match a given set of criteria.
+Automatically retesting PRs in the pool and merging them if they pass tests.
```
Comment (cjwagner) [inline prow/tide/README.md]: ... if they pass tests against the latest base branch commit.

File: prow/tide/README.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  - repo:                   // org/repo
+    labels:                 // List of must have labels
+    missingLabels:          // List of can't have labels
+    excludedBranches:       // Ignore branches
+    includedBranches:       // Include-only branches
+    reviewApprovedRequired: // Review approved is must
+  sync_period:            // Sync jobs period
+  status_update_period:   // PR status update period
+  merge_method:           // Squash, rebase or just merge PR
+  target_url:             // ???
```
Comment (cjwagner) [inline prow/tide/README.md]: This is the URL that is linked to from the `details` link on the Github status context.

File: prow/tide/README.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+* `sync_period`: The field specifies how often Tide will sync jobs with Github. Defaults to 1m.
+* `status_update_period`: The field specifies how often Tide will update Github status contexts.
+   Defaults to the value of `sync_period`.
+* `queries`: List of queries (described below).
+* `merge_method`: A key/value pair of an `org/repo` as the key and merge method to override
+   the default method of merge as value. Valid options are `squash`, `rebase`, and `merge`.
+   Defaults to `merge`.
+* `target_url`: URL for tide status contexts.
+* `pr_status_base_url`: The base URL for the PR status page.
```
Comment (cjwagner) [inline prow/tide/README.md]: Please also note that, if specified, this URL is used to construct a link that will be used for the tide status context so it is mutually exclusive with the `target_url` option.

File: prow/tide/README.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+* `excludedBranches`: List of branches that get excluded when quering the `repos`.
+* `includedBranches`: List of branches that get included when quering the `repos`.
+* `reviewApprovedRequired`: If set, each PR in the query must have review approved.
+
+Under the hood, a query constructed from the fields follows rules described in
+https://help.github.com/articles/searching-issues-and-pull-requests/
+
+**Important**: Each query must return a different set of PRs. No two queries are allowed to contain the same PR.
+
+Every PR in any of the pools that is not **mergeable** is discarded from processing.
```
Comment (cjwagner) [inline prow/tide/README.md]: Please clarify this to specify that this means we discard PRs that require a rebase ("mergeable" doesn't mean meeting merge requirements here).
Maybe just say "PRs that need to be rebased are filtered from the pool before processing."

File: prow/tide/README.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+### Queries
+
+The `queries` field specifies a list of queries.
+Each query corresponds to a set of **open** PRs as candidates for merging.
+It can consist of the following dictionary of fields:
+
+* `repos`: List of queried repositories.
+* `labels`: List of labels any given PR from the `repos` must posses.
+* `missingLabels`: List of labels any given PR from the `repos` must not posses.
```
Comment (cjwagner) [inline prow/tide/README.md]: Drop the mention of `repos` from this line and above. It isn't quite correct because we are actually considering PRs from the `repos` **that are in an appropriate branch** as determined by the branch parameters.

It should be obvious enough that a PR must satisfy all the requirements in this list to be part of the query output.

File: prow/tide/README.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+  * **Specification of PR tests**:
+    A set of jobs that are run over each PR before it gets merged.
+    No job is required to be provided in which case it's enough a PR has all
+    relevant labels set. If a set of jobs is specified,
+    all must succeed (unless configured otherwise).
+    Configured under `presubmits` key of the configuration file.
+
+## PR pools
+
+The set of criteria is specified through the following collection of items:
```
Comment (fejta) [inline prow/tide/README.md]: Consider linking to something like https://github.com/kubernetes/test-infra/blob/master/prow/config/tide.go#L33

File: prow/tide/README.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+* `repos`: List of queried repositories.
+* `labels`: List of labels any given PR from the `repos` must posses.
+* `missingLabels`: List of labels any given PR from the `repos` must not posses.
+* `excludedBranches`: List of branches that get excluded when quering the `repos`.
+* `includedBranches`: List of branches that get included when quering the `repos`.
+* `reviewApprovedRequired`: If set, each PR in the query must have review approved.
+
+Under the hood, a query constructed from the fields follows rules described in
+https://help.github.com/articles/searching-issues-and-pull-requests/
```
Comment (cjwagner) [inline prow/tide/README.md]: "... Therefore every query is really just a structured definition of a standard GitHub search query which can be used to list mergeable PRs."

To make this even more obvious we could annotate the above list of fields with examples of the corresponding search tokens.
`repos`: `repo:org/repo`
`labels`: `label:lgtm`
`missingLabels`: `-label:do-not-merge`
`excludedBranches`: `-branch:dev`
`includedBranches`: `branch:master`
`reviewApprovedRequired`: `review:approved`

File: prow/tide/README.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+**Important**: Each query must return a different set of PRs. No two queries are allowed to contain the same PR.
```
Comment (fejta) [inline prow/tide/README.md]: Is this true? It doesn't seem like it should be

File: prow/tide/README.md [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+**Explanation**: The component starts periodically querying all PRs in `github.com/kubeflow/community` and
+`github.com/kubeflow/examples` repositories that have `lgtm` and `approved` labels set
+and do not have `do-not-merge`, `do-not-merge/hold`, `do-not-merge/work-in-progress`, `needs-ok-to-test` and `needs-rebase` labels set.
+All PRs that conform to the criteria are processed and merged.
+The processing itself can include running jobs (e.g. tests) to verify the PRs are good to go.
+At the same time all commits in PRs from `github.com/kubeflow/community` repository are squashed before merging.
+
+## Presubmits
+
+Before a PR is merged, Tide runs a set of jobs configured in `presubmits` part of the `config.yaml` file to verify the PR can be merged.
```
Comment (cjwagner) [inline prow/tide/README.md]: Please change 
>Tide runs a set of jobs configured in `presubmits` part of the `config.yaml` file to verify the PR can be merged.

To: "Tide ensures that all jobs configured as required in the `presubmits` part of the `config.yaml` file are passing against the latest base branch commit, rerunning the jobs if necessary."

File: prow/tide/README.md [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+Semantic of individual fields of the `presubmits` is described in [prow/README.md#how-to-add-new-jobs](https://github.com/kubernetes/test-infra/blob/master/prow/README.md#how-to-add-new-jobs).
+
+The `presubmits` part consists of a set of `org/repo` based jobs. A PR is merged only if all jobs that are specified for PR's repository succeed (unless configured otherwise).
```
Comment (cjwagner) [inline prow/tide/README.md]: This can probably be removed, it has been mentioned above.

File: prow/tide/README.md [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+### Example
```
Comment (cjwagner) [inline prow/tide/README.md]: This section is outside the scope of Tide. I think you can delete it. The link you provided above about adding new jobs should be sufficient.

File: prow/tide/README.md (hunk 9 above)
Comment (cjwagner) [reply prow/tide/README.md]: It is true, but it probably shouldn't be. We don't de-dup PRs from multiple queries, but we totally could.

Comment (cjwagner) [reply prow/tide/README.md]: Fix is part of #7961

File: prow/tide/README.md (hunk 1 above)
Comment (ingvagabund) [reply prow/tide/README.md]: Done

File: prow/tide/README.md (hunk 2 above)
Comment (ingvagabund) [reply prow/tide/README.md]: Done

File: prow/tide/README.md (hunk 3 above)
Comment (ingvagabund) [reply prow/tide/README.md]: Done

File: prow/tide/README.md (hunk 4 above)
Comment (ingvagabund) [reply prow/tide/README.md]: Done

File: prow/tide/README.md (hunk 5 above)
Comment (ingvagabund) [reply prow/tide/README.md]: Done

File: prow/tide/README.md (hunk 6 above)
Comment (ingvagabund) [reply prow/tide/README.md]: Done

File: prow/tide/README.md (hunk 8 above)
Comment (ingvagabund) [reply prow/tide/README.md]: Done

File: prow/tide/README.md (hunk 10 above)
Comment (ingvagabund) [reply prow/tide/README.md]: Done

File: prow/tide/README.md (hunk 11 above)
Comment (ingvagabund) [reply prow/tide/README.md]: Done

File: prow/tide/README.md (hunk 12 above)
Comment (ingvagabund) [reply prow/tide/README.md]: Done

File: prow/tide/README.md (hunk 7 above)
Comment (ingvagabund) [reply prow/tide/README.md]: Done

File: prow/tide/README.md (hunk 9 above)
Comment (ingvagabund) [reply prow/tide/README.md]: I left the sentence as it is. It can be change later once #7961 is merged.

**Comment by ingvagabund** [comment]
@cjwagner @fejta thanks for the review. I will squash the commits once the review is done. Unless already configured in the `tide.conf` for this repo :P

File: prow/tide/README.md [hunk 13]
```diff
@@ -0,0 +1,131 @@
+# Tide
+
+Tide is a [Prow](https://github.com/kubernetes/test-infra/blob/master/prow/README.md)
+component for managing a pool of PRs that match a given set of criteria.
+Automatically retesting PRs in the pool and merging them if they pass tests
```
Comment (kargakis) [inline prow/tide/README.md]: s/Automatically retesting PRs in the pool and merging/It will automatically retest and merge PRs in the pool/

File: prow/tide/README.md [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+file. It consists of two parts:
+
+  * **Specification of PR pools/criteria**:
+    A set of all PRs Tide keeps track of.
+    The PRs are periodically checked if they are ready to be merged.
+    Configured under `tide` key of the configuration file.
+
+  * **Specification of PR tests**:
+    A set of jobs that are run over each PR before it gets merged.
+    No job is required to be provided in which case it's enough a PR has all
```
Comment (kargakis) [inline prow/tide/README.md]: it's enough if

File: prow/tide/README.md [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+    relevant labels set. If a set of jobs is specified,
+    all must succeed (unless configured otherwise).
```
Comment (kargakis) [inline prow/tide/README.md]: the "unless configured otherwise" part is still not merged, right? Today, every commit status in a PR needs to be green for tide to merge the PR.

File: prow/tide/README.md (hunk 13 above)
Comment (ingvagabund) [reply prow/tide/README.md]: I am fan of the first form :) It sounds more poetic. Though, the latter is fine too :).

File: prow/tide/README.md [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+  queries:                // List of queries
+  - repo:                   // org/repo
+    labels:                 // List of must have labels
+    missingLabels:          // List of can't have labels
+    excludedBranches:       // Ignore branches
+    includedBranches:       // Include-only branches
+    reviewApprovedRequired: // Review approved is must
+  sync_period:            // Sync jobs period
+  status_update_period:   // PR status update period
+  merge_method:           // Squash, rebase or just merge PR
```
Comment (kargakis) [inline prow/tide/README.md]: Is any normalization happening during startup or does this need to be "squash" instead of "Squash"

File: prow/tide/README.md [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+  target_url:             // URL linked to from the details link on the Github status context
+  pr_status_base_url:     // PR status page
+  max_goroutines:         // Per-pool parallelism
+```
+
+Depending on your criteria, some of the items may by omitted.
+
+See [prow/config/tide.go](https://github.com/kubernetes/test-infra/blob/master/prow/config/tide.go)
```
Comment (kargakis) [inline prow/tide/README.md]: I know that @fejta asked for this but isn't it more friendly to point to godoc instead of the source code?

https://godoc.org/github.com/kubernetes/test-infra/prow/config#Tide

File: prow/tide/README.md (hunk 15 above)
Comment (ingvagabund) [reply prow/tide/README.md]: What about the `skip_report` field then? Or the `Tide` does not care about the PR status and checks the Prow jobs directly?

File: prow/tide/README.md (hunk 17 above)
Comment (ingvagabund) [reply prow/tide/README.md]: This is what I was looking for. Thanks!

File: prow/tide/README.md [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+**Explanation**: The component starts periodically querying all PRs in `github.com/kubeflow/community` and
+`github.com/kubeflow/examples` repositories that have `lgtm` and `approved` labels set
+and do not have `do-not-merge`, `do-not-merge/hold`, `do-not-merge/work-in-progress`, `needs-ok-to-test` and `needs-rebase` labels set.
+All PRs that conform to the criteria are processed and merged.
+The processing itself can include running jobs (e.g. tests) to verify the PRs are good to go.
+At the same time all commits in PRs from `github.com/kubeflow/community` repository are squashed before merging.
+
+## Presubmits
+
+Before a PR is merged, Tide ensures that all jobs configured as required in the `presubmits` part of the `config.yaml` file
```
Comment (kargakis) [inline prow/tide/README.md]: Explain how a job is set as required (`skip_report`: `true`)

File: prow/tide/README.md (hunk 15 above)
Comment (kargakis) [reply prow/tide/README.md]: Yeah, skip_report works fine. Please document it.

File: prow/tide/README.md (hunk 18 above)
Comment (kargakis) [reply prow/tide/README.md]: > Explain how a job is set as required (skip_report: true)

Sorry, skip_report=false makes it required.

File: prow/tide/README.md (hunk 16 above)
Comment (ingvagabund) [reply prow/tide/README.md]: I will rephrase it as `Set to squash, rebase or just merge PR`

File: prow/tide/README.md (hunk 18 above)
Comment (ingvagabund) [reply prow/tide/README.md]: Raither `skip_report: false`?

File: prow/tide/README.md [hunk 19]
```diff
@@ -125,7 +124,7 @@ At the same time all commits in PRs from `github.com/kubeflow/community` reposit
 
 ## Presubmits
 
-Before a PR is merged, Tide ensures that all jobs configured as required in the `presubmits` part of the `config.yaml` file
+Before a PR is merged, Tide ensures that all jobs configured as required (skip_report: true) in the `presubmits` part of the `config.yaml` file
```
Comment (kargakis) [inline prow/tide/README.md]: skip_report: false

**Comment by kargakis** [comment]
/lgtm
/hold cancel

because I need this yesterday

**Comment by BenTheElder** [comment]
Literal symlinks browse pretty badly on GitHub, FWIW.

**Comment by MHBauer** [comment]
Place a hypertext link in the prow README to the tide README.
